text/0432-contextual-helpers.md

```diff
@@ -0,0 +1,811 @@
+
+      Change it:
+      {{input value=this.firstName}}
+      {{input value=this.lastName}}
+
+      <button {{action "submit"}}></button>
+      {{!-- end tempalte content ---}}
+    {{/let}}
+  {{/let}}
+{{/let}}
+```
+
+While this largely matches how things work today, there are a few notable
+differences where globals behave "unexpectedly" in this world.
+
+First of all, it is not possible to reference a component, helper or modifier
+in templates without invoking them today:
+
+```hbs
+{{!-- if `join-words` is a global helper, this works as expected --}}
+{{!-- this invokes the helper and yield the result --}}
+
+{{yield (join-words "foo" "bar" separator=",")}}
+         ~~~~~~~~~~
+
+{{!-- however, in this position, Ember does not "see" the helper --}}
+{{!-- this falls back to looking up the `join-words` property on `this` --}}
+
+{{yield join-words}}
+        ~~~~~~~~~~
+
+{{!-- as opposed to a "true" variable/value binding... --}}
+{{!-- this yields the helper as a value, as expected --}}
+
+{{#let (helper "join-words") as |join-words|}}
+  {{yield join-words}}
+          ~~~~~~~~~~
+{{/let}}
+```
+
+In the long term, we poropose to unify the semantics such that globals will
+behave exactly like local bindings (i.e. we should make this second case work).
+
+However, is not possible in the short term. This is due to the ambiguity
+between referencing a global variable and the
+[property lookup fallback](https://github.com/emberjs/rfcs/blob/master/text/0308-deprecate-property-lookup-fallback.md)
+feature. We propose we simply wait until the property lookup fallback is fully
+deprecated and removed, at which point we can reclaim the syntax.
+
+In the mean time, globals can be referenced explicitly using the `component`,
+`helper`, and `modifier` helpers.
+
+Another difference is how global helpers can be invoked without arguments in
+named arguments position today:
+
+```hbs
+{{!-- if `pi` is a global helper that returns the value of the constant 𝛑 --}}
+{{!-- this invokes the helper and passes the value 3.1415... --}}
+<MyComponent @value={{pi}} />
+                      ~~
+
+{{!-- as opposed to a "true" variable/value binding... --}}
+
+{{#let (helper "pi") as |pi|}}
+  {{!-- this passes the helper into the component --}}
+  <MyComponent @value={{pi}} />
+                        ~~
+
+  {{!-- this invokes the helper and passes the value 3.1415... --}}
+  <MyComponent @value={{(pi)}} />
+                         ~~
+{{/let}}
+```
+
+Notably, this problem only exists in the named arguments position. For content
+and attribute positions, it would not makae sense to pass a helper "by value",
+so the ambiguity does not exist (so it always invokes). For sub-expression
+positions (which includes argument positions for curly invocations), the
+parentheses are already mandatory (otherwise it invokes the property fallback).
+
+We propose to deprecate invoking global helpers in named argument positions and
```

Comment:
Is is this deprecation part of this RFC? Does it need its own transition path (ala https://emberjs.github.io/rfcs/0308-deprecate-property-lookup-fallback.html#transition-path)?

Reply:
It is part of this RFC. It should say "deprecate auto-invoking global helpers with no arguments in named arguments positions" and the transition path is from `@foo={{bar}}` to `@foo={{(bar)}}`.